### PR TITLE
CON-1668 : Add support to deploy cv-csp using helm and operator

### PR DESCRIPTION
**Pull request by ckallianpur**
Signed-off-by: chaitra <chaitra.kallianpur@hpe.com>

### CODE REVIEW

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    metadata:
+      labels:
+        app: cv-csp
+    spec: 
+      priorityClassName: system-cluster-critical
+      serviceAccount: hpe-csi-controller-sa
+      containers:
+        - name: cv-csp
+          {{- if .Values.registry }}
+          image: {{ .Values.registry }}/hpestorage/cv-csp:latest
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/templates/cv-csp.yaml:23]: This should reflect the same version as the default below.

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          {{- else }}
+          image: quay.io/hpestorage/cv-csp:v1.0.0
+          {{- end }}
+          imagePullPolicy: {{ .Values.imagePullPolicy | quote }}
+          env:
+          - name: API_HOST
+            value: "https://demo.cloudvolumes.hpe.com"
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/templates/cv-csp.yaml:30]: This should point to prod.

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
+---
+### CSP deployment ###
+kind: Deployment
+apiVersion: apps/v1
+metadata:
+  name: csp-service
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/templates/cv-csp.yaml:6]: cv-csp

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            value: "https://demo.cloudvolumes.hpe.com"
+          ports:
+          - containerPort: 8080
+
+---
+### CSP Service ###
+kind: Service
+apiVersion: v1
+metadata:
+  name: cloudvol-csp-svc
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/templates/cv-csp.yaml:39]: cv-csp-svc

**Review by datamattsson — CHANGES_REQUESTED** [review]
Could you copy the Nimble template and `s/nimble/cv/` and add any CV specific things, like API_HOST

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml (hunk 1 above)
Comment (ckallianpur) [reply helm/charts/hpe-csi-driver/templates/cv-csp.yaml]: done

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml (hunk 2 above)
Comment (ckallianpur) [reply helm/charts/hpe-csi-driver/templates/cv-csp.yaml]: done

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml (hunk 3 above)
Comment (ckallianpur) [reply helm/charts/hpe-csi-driver/templates/cv-csp.yaml]: changed to cv-csp

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml (hunk 4 above)
Comment (ckallianpur) [reply helm/charts/hpe-csi-driver/templates/cv-csp.yaml]: changed to cv-csp-svc

**Comment by datamattsson** [comment]
I rather we kept the formatting of the CSP templates rather intact between the different ones. Here's the copied and pasted with your changes.
```
---
### CSP Service ###
kind: Service
apiVersion: v1
metadata:
  name: cv-csp-svc
  namespace: {{ .Release.Namespace }}
  labels:
    app: cv-csp-svc
spec:
  ports:
  - port: 8080
    protocol: TCP
  selector:
    app: cv-csp

---
### CSP deployment ###
kind: Deployment
apiVersion: apps/v1
metadata:
  name: cv-csp
  namespace: {{ .Release.Namespace }}
spec:
  selector:
    matchLabels:
      app: cv-csp
  replicas: 1
  template:
    metadata:
      labels:
        app: cv-csp
    spec:
      serviceAccount: hpe-csp-sa
      {{- if and (eq .Capabilities.KubeVersion.Major "1") ( ge  ( trimSuffix "+" .Capabilities.KubeVersion.Minor ) "17") }}
      priorityClassName: system-cluster-critical
      {{- end }}
      containers:
        - name: cv-csp
          {{- if .Values.registry }}
          image: {{ .Values.registry }}/hpestorage/cv-csp:v1.0.0
          {{- else }}
          image: quay.io/hpestorage/cv-csp:v1.0.0
          {{- end }}
          imagePullPolicy: {{ .Values.imagePullPolicy | quote }}
          ports:
            - containerPort: 8080
          env:
            - name: API_HOST
              value: "https://prod.cloudvolumes.hpe.com"
      tolerations:
        - effect: NoExecute
          key: node.kubernetes.io/not-ready
          operator: Exists
          tolerationSeconds: 30
        - effect: NoExecute
          key: node.kubernetes.io/unreachable
          operator: Exists
          tolerationSeconds: 30
```

Could you also add `cv-csp.yaml` (copied in the same way as for Helm) here: https://github.com/hpe-storage/co-deployments/tree/master/yaml/csi-driver/edge for this PR?

File: yaml/csi-driver/edge/cv-csp.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  template:
+    metadata:
+      labels:
+        app: cv-csp
+    spec:
+      priorityClassName: system-cluster-critical
+      serviceAccount: hpe-csi-controller-sa
+      containers:
+        - name: cv-csp
+          image: quay.io/hpestorage/cv-csp:latest
```
Comment (datamattsson) [inline yaml/csi-driver/edge/cv-csp.yaml:30]: we use :edge as tag for edge releases.

**Review by datamattsson — CHANGES_REQUESTED** [review]
Could you please copy & paste the nimble YAML declarations and decorate with CV specific names, images and requirements?

File: yaml/csi-driver/edge/cv-csp.yaml (hunk 5 above)
Comment (ckallianpur) [reply yaml/csi-driver/edge/cv-csp.yaml]: made the change

**Comment by ckallianpur** [comment]
> Could you please copy & paste the nimble YAML declarations and decorate with CV specific names, images and requirements?

It might not work. For instance, the nimble csp knows how to write to log files but the cv-csp doesnt know that.

**Comment by ckallianpur** [comment]
> > Could you please copy & paste the nimble YAML declarations and decorate with CV specific names, images and requirements?
> 
> It might not work. For instance, the nimble csp knows how to write to log files but the cv-csp doesnt know that.

I have made the suggested changes

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+          - containerPort: 8080
+      tolerations:
+        - effect: NoExecute
+          key: node.kubernetes.io/not-ready
+          operator: Exists
+          tolerationSeconds: 30
+        - effect: NoExecute
+          key: node.kubernetes.io/unreachable
+          operator: Exists
+          tolerationSeconds: 30
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/templates/cv-csp.yaml:60]: Could you add a line break here. Certain linters don't like this.

**Review by datamattsson — APPROVED** [review]
Small nit and LGTM

**Review by datamattsson — APPROVED** [review]
LGTM

File: helm/charts/hpe-csi-driver/templates/cv-csp.yaml (hunk 2 above)
Comment (datamattsson) [reply helm/charts/hpe-csi-driver/templates/cv-csp.yaml]: Shouldn't this be called out in the secret?

Comment (ckallianpur) [reply helm/charts/hpe-csi-driver/templates/cv-csp.yaml]: Created a defect to move this later

File: yaml/csi-driver/edge/cv-csp.yaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  template:
+    metadata:
+      labels:
+        app: cv-csp
+    spec:
+      serviceAccount: hpe-csp-sa
+      priorityClassName: system-cluster-critical
+      containers:
+        - name: cv-csp
+          image: quay.io/hpestorage/cv-csp:edge
```
Comment (raunakkumar) [inline yaml/csi-driver/edge/cv-csp.yaml:51]: nit: we can update his to 1.5.0-beta once the helm chart is pushed.

**Review by raunakkumar — APPROVED** [review]

**Pull request merged**